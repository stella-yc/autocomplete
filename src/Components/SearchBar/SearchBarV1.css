.Searchbar-component-v1 {
  color: black;
  font-family: 'Montserrat', sans-serif;
  border: none;
}

.Searchbar-container-v1 {
  height: 2.5em;
  width: 90%;
  max-width: 540px;
  margin: 0 auto;
  border: none;
  border-radius: 50px;
  box-shadow: 0px 3px 20px 4px rgba(53, 53, 53, 0.1);
  background-color: white;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  overflow: hidden;
}

.Searchbar-input-v1 {
  height: 100%;
  width: 80%;
  margin-left: 2em;
  border: none;
  font-size: 1.1em;
  font-family: 'Montserrat', sans-serif;
}

.Searchbar-input-v1:focus {
  outline: none;
}

.Searchbar-icon-v1 {
  width: 30%;
  line-height: 2.5em;
  border-radius: 20px 0 0 20px;
  border: none;
  color: white;
  background-color: #53a63a;
  cursor: pointer;
}

.Searchbar-results-v1 {
  width: 90%;
  max-width: 540px;
  margin: 1em auto;
  padding: 0;
  border: none;
  border-radius: 20px;
  box-shadow: 0px 3px 20px 4px rgba(53, 53, 53, 0.1);
  background-color: white;
  list-style-type: none;
  text-align: left;
  overflow: hidden;
}
