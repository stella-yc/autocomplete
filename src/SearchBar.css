.Searchbar-container {
  color: black;
}

.Searchbar-input {
  width: 80%;
  height: 2em;
  padding: 0 0.5em;
}

.Searchbar-results {
  list-style-type: none;
  text-align: left;
}

.SearchResult-selected {
  background-color: pink;
}
