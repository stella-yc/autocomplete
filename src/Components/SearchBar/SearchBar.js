import React, { Component } from 'react';
import PropTypes from 'prop-types'

import SearchResult from '../SearchResult';
import './SearchBarV1.css';
import './SearchBarV2.css';

class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: 0,
      typedValue: '',
      results: [],
      navigateToSelected: false
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.retrieveResults = this.retrieveResults.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.visitSearchResult = this.visitSearchResult.bind(this);
  }

  handleInputChange (event) {
    const typedValue = event.target.value;
    this.setState({ typedValue }, () => this.retrieveResults());
  }

  retrieveResults () {
    this.props.retrieveSearchResults(this.state.typedValue)
      .then(results => this.setState({ results }));
  }

  handleKeyDown (event) {
    const { selected, results } = this.state;
    // if up arrow key pressed
    if (event.keyCode === 38 && selected > 0) {
      this.setState((prevState =>
        ({ selected: prevState.selected - 1 })
      ));
    }
    // if down arrow key pressed
    if (event.keyCode === 40 && selected <= results.length) {
      this.setState((prevState =>
      ({ selected: prevState.selected + 1 })
    ));
    }
    // if enter key is pressed
    if (event.keyCode === 13) {
      this.visitSearchResult();
    }
  }

  visitSearchResult () {
    if (this.state.selected === 0) {
      this.setState({
        navigateToSelected: true,
        selected: 1
      });
    } else {
      this.setState({navigateToSelected: true});
    }
  }

  render () {
    const { selected, typedValue, results, navigateToSelected } = this.state;
    let { style } = this.props;
    if (!style) {
      style = 'v1';
    }

    return (
      <div className={`Searchbar-component-${style}`}>
        <div className={`Searchbar-container-${style}`}>
          <input
            className={`Searchbar-input-${style}`}
            type="text"
            placeholder="Enter a college or university..."
            onChange={this.handleInputChange}
            value={typedValue}
            onKeyDown={this.handleKeyDown}
          />
          <span
            className={`Searchbar-icon-${style}`}
            onClick={this.visitSearchResult}
          >
            <i className="fa fa-search" aria-hidden="true"></i>
          </span>
        </div>
        {
          !results.length
            ? null
            : <ul className={`Searchbar-results-${style}`}>
              {
                results.map((result, idx) =>
                  <SearchResult
                    style={style}
                    goToLink={navigateToSelected}
                    selected={selected === (idx + 1)}
                    key={result.id}
                    result={result}
                    typedValue={typedValue}
                  />
                )
              }
            </ul>
        }
      </div>
    );
  }
}

export default SearchBar;

SearchBar.propTypes = {
  style: PropTypes.oneOf(['v1', 'v2']),
  retrieveSearchResults: PropTypes.func.isRequired
};
