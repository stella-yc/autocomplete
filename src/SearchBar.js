import React, { Component } from 'react';
import './SearchBar.css';

import promiseJSONP from './utils/JSONPUtil';
import generateSearchUrl from './utils/searchUrl';
import { firstTenElements, uniqueVals } from './utils/arrayUtil';
import SearchResult from './SearchResult';

class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: 0,
      typedValue: '',
      results: [],
      navigateToSelected: false
    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.retrieveResults = this.retrieveResults.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  handleInputChange (event) {
    const typedValue = event.target.value;
    this.setState({ typedValue }, () => this.retrieveResults());
  }

  retrieveResults () {
    const api = generateSearchUrl(this.state.typedValue);
    promiseJSONP(api)
      .then(data => {
        const uniqueResults = uniqueVals(data.results)
        const results = firstTenElements(uniqueResults);
        this.setState({results});
      })
      .catch(err => console.log(err));
  }

  handleKeyDown (event) {
    const { selected, results } = this.state;
    // if up arrow key pressed
    if (event.keyCode === 38 && selected > 0) {
      this.setState((prevState =>
        ({ selected: prevState.selected - 1 })
      ));
    }
    // if down arrow key pressed
    if (event.keyCode === 40 && selected <= results.length) {
      this.setState((prevState =>
      ({ selected: prevState.selected + 1 })
    ));
    }
    // if enter key is pressed
    if (event.keyCode === 13) {
      this.setState({navigateToSelected: true});
    }
  }

  render () {
    const { selected, typedValue, results, navigateToSelected } = this.state;
    return (
      <div className="Searchbar-container">
        <input
          className="Searchbar-input"
          type="text"
          placeholder="Enter a college or university..."
          onChange={this.handleInputChange}
          value={typedValue}
          onKeyDown={this.handleKeyDown}
        />
        <ul className="Searchbar-results">
          {
            results.map((result, idx) =>
              <SearchResult
                goToLink={navigateToSelected}
                selected={selected === (idx + 1)}
                key={result.id}
                result={result}
                typedValue={typedValue}
              />
            )
          }
        </ul>
      </div>
    );
  }
}

export default SearchBar;
