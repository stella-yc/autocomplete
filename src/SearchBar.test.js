import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import SearchBar from './SearchBar';
import renderer from 'react-test-renderer';

it('renders shallowly without crashing', () => {
  shallow(<SearchBar />);
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<SearchBar />, div);
});

it('renders correctly', () => {
  const tree = renderer.create(
    <SearchBar />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('updates state when user types into input', () => {});

it('queries api when user types into input', () => {});

it('updates state when up arrow key is pressed on input', () => {});

it('updates state when up down key is pressed on input', () => {});

it('updates state when up enter key is pressed on input', () => {});

it('if no style is provided, default to v1', () => {});

it('if supplied with style, classNames reflect style', () => {});

it('if there are no search results, ul element is absent', () => {});

it('if there are search results, SearchResult component(s) is present', () => {});
