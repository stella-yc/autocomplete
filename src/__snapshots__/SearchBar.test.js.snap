// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> Component renders correctly 1`] = `
<div
  className="Searchbar-component-v1"
>
  <div
    className="Searchbar-container-v1"
  >
    <input
      className="Searchbar-input-v1"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter a college or university..."
      type="text"
      value=""
    />
    <span
      className="Searchbar-icon-v1"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fa fa-search"
      />
    </span>
  </div>
</div>
`;
